### PR TITLE
Add FieldMetadata to StructField

### PR DESCRIPTION
**Pull request by allisonport-db**
Currently the DSR ignores column metadata when reading the schema. This PR introduces a new type `FieldMetadata` which stores the metadata for a given field within the schema.
Changes:
• Add metadata attribute to `StructField`
• Adds a `FieldMetadata` class
• Includes column metadata in `DataTypeParser` `toJson` and `fromJson`
• New tests for schemas with column metadata

### CODE REVIEW

File: standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala [hunk 1]
```diff
@@ -131,11 +136,12 @@ private[standalone] object DataTypeParser {
...
   private def parseStructField(json: JValue): StructField = json match {
     case JSortedObject(
-    ("metadata", _: JObject),
+    ("metadata", metadata: JObject),
     ("name", JString(name)),
     ("nullable", JBool(nullable)),
     ("type", dataType: JValue)) =>
-      new StructField(name, parseDataType(dataType), nullable)
+      new StructField(name, parseDataType(dataType), nullable,
+         metadata.obj.toMap[String, AnyRef].asJava)
```
Comment (allisonport-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala:144]: This stores the metadata values inn StructField as a JValue, is that what we want?

File: standalone/src/main/java/io/delta/standalone/types/StructField.java [hunk 2]
```diff
@@ -57,6 +61,20 @@ public StructField(String name, DataType dataType, boolean nullable) {
...
         this.dataType = dataType;
         this.nullable = nullable;
+        this.metadata = Collections.emptyMap();
+    }
+
+    /**
+     * @param name  the name of this field
+     * @param dataType  the data type of this field
+     * @param nullable  indicates if values of this field can be {@code null} values
+     * @param metadata metadata for this field
```
Comment (scottsand-db) [inline standalone/src/main/java/io/delta/standalone/types/StructField.java:71]: nit: add another space

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaDataReaderSuite.scala [hunk 3]
```diff
@@ -336,4 +336,15 @@ class DeltaDataReaderSuite extends FunSuite {
       new IntegerType,
       new ArrayType(new DoubleType, true),
       false))
+
+  test("toJson fromJson with metadata") {
+    val schema_str =
+      """{"type":"struct",
+        |"fields":[
+        |{"name":"a","type":"integer","nullable":true,"metadata":{"test":"foo"}},
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaDataReaderSuite.scala:344]: you should also test an empty metadata here

**Review by scottsand-db — CHANGES_REQUESTED** [review]
By biggest question, and this will help answer your question of "should we store the metadata map values in StructField as a JValue", is why aren't we including a `FieldMetadataBuilder.java` and `FieldMetadata.java` like [spark/.../Metadata](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala) does?(renamed to put the `Field` keyword in there to avoid confusion with `actions/Metadata.java`)

Since we are exposing a `Map<String, Object>`, users can put an entry `("foo", new Foo())`. This would break things during serialization, and we want to prevent that. Limiting the column metadata value types would help prevent this.

I think this should look more like [spark/.../Metadata](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala).

@zsxwing what do you think?

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaDataReaderSuite.scala [hunk 4]
```diff
@@ -336,4 +336,15 @@ class DeltaDataReaderSuite extends FunSuite {
       new IntegerType,
       new ArrayType(new DoubleType, true),
       false))
+
+  test("toJson fromJson with metadata") {
+    val schema_str =
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaDataReaderSuite.scala:341]: super small nit: schemaStr instead

File: standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala (hunk 1 above)
Comment (zsxwing) [reply standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala:144]: We should not expose JValue in public APIs. Could you convert them to java objects instead?

**Comment by zsxwing** [comment]
> Since we are exposing a `Map<String, Object>`, users can put an entry `("foo", new Foo())`. This would break things during serialization, and we want to prevent that. Limiting the column metadata value types would help prevent this.

Yep. This is a good idea.

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+import java.util.Map;
+import java.util.HashMap;
+import java.util.stream.Collectors;
+
+
+/**
+ * The metadata for a given {@link StructField}.
+ */
+public final class FieldMetadata {
+    // TODO: should this be an immutable map created via the builder?
```
Comment (allisonport-db) [inline standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:52]: should this be an immutable map?

File: standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala [hunk 6]
```diff
@@ -111,10 +111,35 @@ private[standalone] object DataTypeParser {
...
+    val metadata = field.getMetadata()
 
     ("name" -> name) ~
       ("type" -> dataTypeToJValue(dataType)) ~
-      ("nullable" -> nullable)
+      ("nullable" -> nullable) ~
+      ("metadata" -> metadataValueToJValue(metadata))
+  }
+
+  def metadataValueToJValue(value: Any): JValue = {
```
Comment (allisonport-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala:122]: Why are some functions in DataTypeParser private and others aren't? Should this one be private?

**Comment by allisonport-db** [comment]
@zsxwing ready for another look

File: standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala (hunk 6 above)
Comment (scottsand-db) [reply standalone/src/main/scala/io/delta/standalone/internal/util/DataTypeParser.scala:122]: Good catch. The only public ones should be `toJson` and `fromJson`; everything else can be private.

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    public static class Builder {
+        private Map<String, Object> metadata =new HashMap<String, Object>();
+
+        public Builder putNull(String key) {
+            metadata.put(key, null);
+            return this;
+        }
+
+        public Builder putLong(String key, Long value) {
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:123]: @zsxwing what do you think?

**Review by scottsand-db — APPROVED** [review]
LGTM with a few minor comments

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (allisonport-db) [reply standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:123]: Makes sense since there's also already `putNull()`. 

@zsxwing @scottsand-db 
What about for the arrays? Is there a situation where an array metadata value might have null as an element?

[the remark this answers is not in the document]
Comment (scottsand-db) [reply standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java]: I would go with whatever Ryan says here. For me, using primitives seems fine.

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+     * @return the value to which the specified key is mapped, or null if there is no mapping for
+     * the given key
+     */
+    public Object get(String key) {
+        return metadata.get(key);
+    }
+
+    @Override
+    public String toString() {
+        return metadata.entrySet()
```
Comment (zsxwing) [inline standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:83]: It's better to add a toJson method and make `toString` return it. Otherwise, this `toString` implementation doesn't work for Array type.

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                e.getValue().equals(that.metadata.get(e.getKey())) ||
+                        (e.getValue().getClass().isArray() &&
+                                that.metadata.get(e.getKey()).getClass().isArray() &&
+                                Arrays.equals(
+                                        (Object[]) e.getValue(),
+                                        (Object[]) that.metadata.get(e.getKey()))));
+    }
+
+    @Override
+    public int hashCode() { return metadata.hashCode(); }
```
Comment (zsxwing) [inline standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:105]: Similar to `equals`, we need to handle Array type here as well. Otherwise, it's possible that equals return true but the two objects have different hash codes.

**Review by zsxwing — COMMENTED** [review]
Left some minor comments. Great work!

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java (hunk 8 above)
Comment (allisonport-db) [reply standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:83]: I updated this so it should work with Array type. The `toJson` method is in `DataTypeParser` is it preferable to reference that from here instead? This function is for `StructField.buildFormattedString`

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                        (e.getValue().getClass().isArray() &&
+                                that.metadata.get(e.getKey()).getClass().isArray() &&
+                                Arrays.equals(
+                                        (Object[]) e.getValue(),
+                                        (Object[]) that.metadata.get(e.getKey()))));
+    }
+
+    @Override
+    public int hashCode() {
+            return metadata.entrySet()
```
Comment (zsxwing) [inline standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:109]: nit: indent

**Review by zsxwing — APPROVED** [review]
LGTM

File: standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java (hunk 10 above)
Comment (scottsand-db) [reply standalone/src/main/java/io/delta/standalone/types/FieldMetadata.java:109]: I'll add investigating Java formatting tests as a ticket, should help both of us with our java code.
